### PR TITLE
feat: add QR-Login / Device Authorization Grant (RFC 8628)

### PR DESCRIPTION
**Pull request by Norrodar**
## Summary

This feature implements the **RFC 8628 Device Authorization Grant** as QR-Login for Pocket ID. Users can sign in on devices without keyboard or WebAuthn support (Smart TVs, consoles, set-top boxes) by scanning a QR code or entering an 8-digit code on a second device.

## Motivation

Pocket ID relies on WebAuthn/Passkeys as primary authentication. Devices like Smart TVs (e.g., LG WebOS) don't support WebAuthn. Without this feature, login on such devices is impossible. RFC 8628 Device Authorization Grant is the standard OAuth flow for exactly this use case.

### Backend

**New Endpoints:**
| Endpoint | Method | Auth | Description |
|----------|--------|------|-------------|
| `/oidc/device/authorize` | POST | Client-Auth | Creates Device Code + User Code |
| `/oidc/device/verify` | POST | User-Auth | Verifies User Code, authorizes device |
| `/oidc/device/info` | GET | User-Auth | Returns scope/client info for a User Code |
| `/oidc/device/exchange-session` | POST | None | Polls for authorization, sets session cookie |
| `/oidc/token` | POST | Client-Auth | Standard token endpoint with `grant_type=urn:ietf:params:oauth:grant-type:device_code` |

**Data Model:**
- `oidc_device_codes` table with `device_code` (UNIQUE), `user_code` (UNIQUE), `scope`, `expires_at`, `is_authorized`, `user_id`, `client_id`, `last_polled_at`, `interval_seconds`
- Device codes: 32 alphanumeric characters (190-bit entropy)
- User codes: 8 characters from `ABCDEFGHJKMNPQRSTUVWXYZ23456789` (no ambiguous characters)

**Self-Login Client:**
- Internal OAuth client `pocket-id-self-login` is auto-created on startup
- Used exclusively for the QR-Login flow on the login page itself
- `/exchange-session` endpoint restricted to this client only

**Security:**
- `FOR UPDATE` locks on all device code operations (race condition protection)
- Rate limiting via `slow_down` with interval increase (+5s per violation, max 60s)
- Device codes deleted immediately after single use
- Expired codes cleaned up by scheduled job
- `Cache-Control: no-store` + `Pragma: no-cache` on all token responses
- Already authorized codes cannot be re-verified or queried
- Admin toggle `qrLoginEnabled` blocks all device flow endpoints

### Frontend (SvelteKit / Svelte 5 / TypeScript)

**New Components:**
- `qr-login-flow.svelte` — QR code display + countdown + polling loop
- `code-input.svelte` — 8-digit code input with auto-focus, paste support, click-to-copy
- `device-verification-modal.svelte` — Modal for code entry in account settings

**New Pages/Changes:**
- `/login/alternative` — Alternative login page with QR code and code input
- `/device` — Device verification page (enter code → confirm scopes → authorize)
- `/login` — Auto-redirect to `/login/alternative` when QR login is active
- `/authorize` — Auto-redirect to `/login/alternative` when WebAuthn unavailable (Smart TVs)
- `/settings/account` — "Connect device" button with device verification modal

**UX Details:**
- Countdown based on absolute `expiresAt` (no timer drift on tab switch)
- `slow_down` handling: reads server interval from response, fallback +5s
- Automatic detection of limited devices (`isLimitedDevice()` utility)
- i18n: German + English

## RFC 8628 Compliance

| Requirement | Status |
|-------------|--------|
| Device Authorization Endpoint (§3.1) | ✅ |
| Device Authorization Response with all required fields (§3.2) | ✅ |
| `verification_uri_complete` (§3.3.1) | ✅ |
| Token Endpoint with `grant_type=urn:ietf:params:oauth:grant-type:device_code` (§3.4) | ✅ |
| Error Codes: `authorization_pending`, `slow_down`, `expired_token`, `access_denied` (§3.5) | ✅ |
| All errors as HTTP 400 (§3.5) | ✅ |
| `slow_down` interval increase +5s (§3.5) | ✅ |
| `slow_down` response includes `interval` field (§3.5) | ✅ |
| `Cache-Control: no-store` on token response (RFC 6749 §5.1) | ✅ |
| Case-insensitive User Code comparison (§6.1) | ✅ |
| Unambiguous character set (§6.1) | ✅ |
| Discovery Metadata: `device_authorization_endpoint` (§4) | ✅ |
| `grant_types_supported` includes Device Code Grant (§4) | ✅ |

## Changed Files (QR-Login specific)

**35 files**, ~1200 lines added, ~300 lines removed.

Backend: `oidc_controller.go`, `oidc_service.go`, `errors.go`, `oidc_dto.go`, `string_util.go`, `well_known_controller.go`, `oidc.go` (model), `app_config.go`, `db_cleanup_job.go`, 4 migrations

Frontend: `qr-login-flow.svelte`, `code-input.svelte`, `device-verification-modal.svelte`, `oidc-service.ts`, `device-detect-util.ts`, `oidc.type.ts`, `login-wrapper.svelte`, `app-config-service.ts`, various pages and i18n files

## Review Process

The code was reviewed in 3 rounds by 4 specialized review agents:
1. **Test Agent** — Bugs, edge cases, race conditions
2. **Code Reviewer** — Security, validations, error handling
3. **RFC 8628 Auditor** — Full protocol compliance
4. **Humanness Reviewer** — Code quality, style consistency

All CRITICAL and HIGH findings were resolved. Remaining LOW findings are deliberate design decisions or affect upstream code.

## Real-World Testing

This feature has been running on my personal Pocket ID instance for ~2 months with 8 family members actively using it. The UX has been iteratively refined based on real usage feedback. Over the last 3-4 weeks, very few changes were needed, which gives me confidence that the feature has reached a stable state.

Beyond organic usage, the feature has been extensively tested through automated review cycles covering security, RFC compliance, edge cases, and code quality.

On a personal note: I love the peace of mind QR-Login has brought — no more "I can't log in on the TV" messages from family 😄

### Qr-Login Page
grafik

### Code-Verification Option
grafik
grafik

### Admin-Toggle
grafik

### UI - Missing CSS-Fallback

Dark mode on Smart TVs — On some Smart TVs (e.g., LG WebOS), text may appear black on a dark background due to outdated browser engines that don't fully support CSS custom properties or `prefers-color-scheme`. The QR code itself remains visible (white container), but surrounding text can be unreadable. This is not specific to the QR-Login feature — the entire UI lacks fallback styles for browsers without CSS custom property support. Fixing this should be addressed in a separate PR.

grafik

### CODE REVIEW

**Comment by Norrodar** [comment]
@ItalyPaleAle @kmendell @team
I need some help for a design decision:

## Context

The simple page (`/simple/`) is a separate, minimal entrypoint for devices without WebAuthn support (e.g. Smart TVs with outdated browser engines). It only supports the Device Authorization Flow: displaying a QR code and a user code so the user can complete the login on another device.

grafik


Since this page is plain static HTML/CSS/JS (no SvelteKit, no bundler), it can't use `import` statements or rely on the build pipeline. The QR code needs to be generated client-side — encoding the verification URL so the user can scan it with their phone.

## QR Code Library for the Simple Page

The main app already uses the [`qrcode`](https://www.npmjs.com/package/qrcode) npm package (MIT license), but it consists of ~30 individual files and requires a bundler — there's no single-file browser build included.

I see two options:

### Option 1: Bundle at build time (in Dockerfile)
Add an `esbuild` step in the Dockerfile that bundles the existing `qrcode` dependency into a single `qrcode.min.js` and copies it into `static/simple/`. Both the main app and the simple page would use the same library — just delivered differently.

### Option 2: Vendor a standalone QR library
Include a pre-built QR code library (e.g. [`qrcode-generator`](https://github.com/kazuhikoarase/qrcode-generator), MIT license) as a static file directly in the repo under `static/simple/`. Simple and no build changes needed, but it's a separate library from what the main app uses and needs manual updates.

My preference would be **Option 1** since it reuses the existing dependency without adding anything to the repo. However, it does add another layer of complexity to the build process. What do you think?

Edit:
I just realized that it makes more sense to put the simple page under `/alternative/simple/` instead. ;)

**Comment by ItalyPaleAle** [comment]
@Norrodar can the server generate the QR code with a new API route?

**Comment by Norrodar** [comment]
@ItalyPaleAle Based on [RFC 8628 Section 3.3.1](https://datatracker.ietf.org/doc/html/rfc8628#section-3.3.1), I assumed that the client must dipslay the URI and may optionally present it as a QR code — so I concluded that generating the QR code is always the clients responsibility. Also all other QR codes are currently generated browser-side. But you're right, for the self-login flow it's a moot point anyway since Pocket ID acts as both server and client.
But yeah — the RFC doesn't actually say anything about who has to generate the QR code, just who **_may present_** it.

To generate QR image server-side, we'd need to add a Go library like [skip2/go-qrcode](https://github.com/skip2/go-qrcode) (MIT) as a new backend dependency. On the upside, this would essentially eliminate all JavaScript dependencies from the simple page. I'd envision the endpoint as: GET /oidc/device/qrcode?data=<url>

I'm going to think about it / work on it as soon as I find some time during the Easter days.
Happy Easter! 🐣

**Comment by ItalyPaleAle** [comment]
Ok I see. I don't think there's anything in the RFC that states that you can't generate in the server, but maybe it's best if the code isn't included in a query string arg.

I'd go with option 1 so all deps are in the package.json

You can create a `build-simple.sh` script that performs the bundling only for the simple page. Add that as a script to the package.json like `build-simple`. I think a simple build with something like esbuild or similar should be fine!

Would that work?

**Comment by kmendell** [comment]
> The main app already uses the [qrcode](https://www.npmjs.com/package/qrcode) npm package (MIT license), but it consists of ~30 individual files and requires a bundler — there's no single-file browser build included.

I dont exactly see what you mean, that's a dependency in the application, and its bundled with the frontend as its already used for the login code link, so its already bundle. -- Unless im completely misunderstanding here (it been a long week..)

File: backend/internal/controller/oidc_controller.go [hunk 1]
```diff
@@ -202,35 +232,41 @@ func (oc *OidcController) createTokensHandler(c *gin.Context) {
...
+	if handleDeviceFlowError(c, err) {
 		return
-	case err != nil:
+	}
+	if err != nil {
 		_ = c.Error(err)
 		return
 	}
 
+	// RFC 6749 Section 5.1: token responses MUST include these cache headers
```
Comment (ItalyPaleAle) [inline backend/internal/controller/oidc_controller.go:329]: There seems to be a lot of unrelated things added to this PR. While I think this is a good example of something we want, please split it into a separate PR

File: backend/internal/utils/string_util.go [hunk 2]
```diff
@@ -17,9 +17,10 @@ func GenerateRandomAlphanumericString(length int) (string, error) {
 	return GenerateRandomString(length, charset)
 }
 
-// GenerateRandomUnambiguousString generates a random string of the given length using unambiguous characters
+// GenerateRandomUnambiguousString generates a random string using unambiguous uppercase characters and digits.
+// Excludes 0/O and 1/I/L to avoid visual confusion.
 func GenerateRandomUnambiguousString(length int) (string, error) {
```
Comment (ItalyPaleAle) [inline backend/internal/utils/string_util.go:22]: This method is also used for codes that are longer-lived (even 1 hour more), and I'm concerned that the new implementation strongly reduces the entropy of the codes.

Maybe we should increase the length of the tokens generated with this method?

File: frontend/static/simple/qr/main.js [hunk 3]
```diff
@@ -0,0 +1,263 @@
+/* Simple login page – no framework, no bundler needed.
```
Comment (ItalyPaleAle) [inline frontend/static/simple/qr/main.js]: This file has direct DOM access everywhere, and I am not sure how maintainable it will be in the future... It gives me jQuery vibes :) Can we use _some_ framework that compiles down to a widely-supported set of browsers? Like, Svelte without SvelteKit?

File: frontend/src/lib/utils/device-detect-util.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
+const LIMITED_DEVICE_PATTERNS = [
```
Comment (ItalyPaleAle) [inline frontend/src/lib/utils/device-detect-util.ts:1]: Please use capabilities detection rather than matching user agents, which is not reliable. Simply detect if WebAuthn is supported, and if not, send to the "simple" page

File: backend/internal/utils/string_util.go (hunk 2 above)
Comment (Norrodar) [reply backend/internal/utils/string_util.go:22]: You're right, reducing the charset affects entropy for longer-lived tokens. Instead of changing the existing function, I'd like to keep GenerateRandomUnambiguousString as it was (with the full mixed-case charset) and add a separate GenerateRandomUnambiguousUppercaseString for the short-lived device codes, where readability on TV screens matters more than maximum entropy. That way, existing token lengths don't need to increase and security stays at the same level.

File: frontend/src/lib/utils/device-detect-util.ts (hunk 4 above)
Comment (Norrodar) [reply frontend/src/lib/utils/device-detect-util.ts:1]: I agree that capability detection is preferable where possible. The redirect logic uses !window.PublicKeyCredential as a capability check for missing WebAuthn support, combined with CSS.supports('color', 'oklch(0 0 0)') to decide between the SvelteKit UI and the simple fallback page.

The UA list is OR'd with the WebAuthn check (isLimitedDevice() || !PublicKeyCredential) to catch devices like Tesla, PlayStation, Xbox, and some Smart TVs that do expose the WebAuthn API but can't practically use passkeys. These are exactly the devices this feature is built for. Without it, those users would land on the passkey login page and may get stuck.

Maintaining the list should be manageable since most entries are model-independent (e.g. car, smarttv etc., also Tesla, webOS, PlayStation). If someone's device passes both capability checks but still can't use passkeys, it's just a one-line addition to the list.

File: frontend/static/simple/qr/main.js (hunk 3 above)
Comment (Norrodar) [reply frontend/static/simple/qr/main.js]: Would need to test whether standalone Svelte (without SvelteKit) stays as minimal. With the simple page and its 5 states, I tried to produce as little overhead as possible — just HTML, CSS, and a bit of JS. I'll give it a try (:

(Sorry for the late answer - I just forgot to press the comment-button :D )

Comment (ItalyPaleAle) [reply frontend/static/simple/qr/main.js]: Svelte's pitch is that it should have as little overhead over direct DOM access as possible. But even if there was some overhead, being able to have a framework (and reuse components with the broader Pocket ID UI) would help a lot with maintainability, long-term.

**Comment by stonith404** [comment]
@Norrodar Thanks for this PR! I was in the military since January until 2 weeks ago and now I have too keep up with all the new issues and PRs. Just to let you know, you PR wasn't forgotten, I'm trying to review it as soon as possible.

**Comment by Norrodar** [comment]
## Small update

Thank you very much for the comments. This project is really important to me — I love how simple it is to set up and use. The QR login was the one real pain point for me, which is why I care so much about getting it right. I know this PR is still a bit messy and unstructured, and I'm doing my best to bring it into good shape.

I've been running this branch in my own production setup at home. That's how most of the bugs fixed in the latest months, and it's also why I kept merging upstream into the branch. But I've realised that this is no substitute for proper tests. I've now moved my home deployment to a separate branch, so this PR can stay quiet — I'll stop pushing upstream merges into it. I'll keep collecting and fixing QR-related bugs, but I'll bundle them into a single update here instead of many small pushes.

I considered splitting this PR into smaller pieces to make the review easier, but for now I've kept it together: most of the change is new, self-contained files rather than edits to existing logic.

In total the PR touches 46 files (~2,000 lines added, ~200 removed). Excluding the lockfile and translation files, roughly 1,900 added / 140 removed are review-relevant. Please let me know if there's anything I can help with.

## Changes overview

I tried to collect all the changes (with some help from AI) to give a clearer overview, grouped by area.

### Backend — self-login client & device flow extensions

- `service/oidc_service.go`
  - new `ensureSelfLoginClient` (creates/maintains the internal `pocket-id-self-login` public client)
  - new `checkDeviceCodePollRate` (RFC 8628 polling interval + `slow_down`)
  - new `ExchangeDeviceTokenForSession` (turns an authorized device code into a browser session)
  - small edits to `CreateDeviceAuthorization` / `VerifyDeviceCode`
- `controller/oidc_controller.go`
  - new `requireQrLogin` gate
  - new `handleDeviceFlowError`
  - new `exchange-session` handler + wiring of the device endpoints
- `utils/string_util.go`
  - new `GenerateRandomUnambiguousUppercaseString` for TV-readable user codes (uppercase, no ambiguous characters)
- `job/db_cleanup_job.go`
  - cleanup job for expired device codes
- migrations `*_device_code_rate_limiting.*`
  - adds `last_polled_at` / `interval_seconds` columns
- config plumbing for a `qrLoginEnabled` admin toggle
  - `model`, `dto`, `app_config_service`, `app-config-general-form.svelte`

### Frontend — SvelteKit QR-login UI
grafik

- `components/qr-login-flow.svelte`
  - QR code + user code display, countdown, polling and session exchange
- `components/code-input.svelte`
  - segmented code input (paste and keyboard navigation, mobile-friendly)
- `settings/account/device-verification-modal.svelte`
  - enter/confirm a code coming from another device
- `utils/device-detect-util.ts`
  - capability + user-agent detection, routing to the right login page (modern vs. simple)
- `utils/redirection-util.ts`
  - shared `isSafeRedirect` helper
- `services/oidc-service.ts` / `types/oidc.type.ts`
  - client API methods and types for the device/self-login endpoints
- route changes
  - `authorize`, `login`, `login/alternative`: redirect limited devices / no-WebAuthn to the QR or alternative login
  - `device`, `settings/account`: code-entry entry points

### Frontend — standalone "simple" page for legacy browsers
A separate, non-SvelteKit bundle (no oklch, no Tailwind, ES2015 target) for browsers that can't render the modern UI.

grafik

- `simple/src/App.svelte`
  - the whole standalone app: device-code request, QR rendering, polling, countdown, all states
- `static/simple/qr/style.css`
  - hand-written CSS, hardcoded hex colors, webkit-prefixed flexbox
- `static/simple/qr/index.html`
  - minimal HTML shell
- build setup
  - `vite.config.ts`, `tsconfig.json`, `main.ts`, `qrcode-entry.js`, `build-simple.sh` (IIFE build, bundles the QR library)

## Review guide

A short description of how I'd go through this PR:

1. **Backend core** — `oidc_service.go` (`ensureSelfLoginClient`, `checkDeviceCodePollRate`, `ExchangeDeviceTokenForSession`) → `oidc_controller.go` (`exchange-session`, `requireQrLogin`). This is the security-relevant heart.
2. **Routing / detection** — `device-detect-util.ts` + `redirection-util.ts`, then the route guards in `authorize/+page.svelte` and `login/+page.svelte`.
3. **SvelteKit UI** — `qr-login-flow.svelte`, `code-input.svelte`, `device-verification-modal.svelte`.
4. **Standalone page** — `simple/src/App.svelte` + `style.css`. Self-contained, own build, no shared imports.

## Bug fixes in the latest push

- **reauth bypass:** a `reauth=1` URL marker used to persist, so a reload could skip required re-authentication. It's now only honored on limited devices and stripped from the URL after use.
- **double submit:** the code modal could fire a second request on a repeated Enter; added a guard.
- **cancel dead-end:** removed a `clearUser()` that left the session unrecoverable when the user cancelled.
- **odd-length separator:** fixed the code-input separator placement for odd lengths.
- **state naming:** renamed a `state` variable that collided with the `$state` rune. This also surfaced and fixed a bug where the scope-confirm view went blank during the request.
- **mobile keyboard:** reworked the code input from one `<input>` per box to a single hidden input behind the boxes, so the on-screen keyboard no longer closes after every character on mobile.

## Testing

Full disclosure: this PR doesn't include automated tests for the new self-login/QR flow yet. I've been running the branch in my own production setup at home, but I know that isn't a substitute for committed tests. I haven't worked much with automated testing in open-source projects so far, but I'd like to learn and add proper tests for this PR. If you have a preference on scope or where they should live, I'll follow it.

## AI usage

In an earlier discussion I mentioned that I'd be using Claude. Just to be transparent: I used an LLM to help with the legacy-page CSS and to review the diff for bugs.

**Comment by JPar99** [comment]
> ## Small update
> Thank you very much for the comments. This project is really important to me — I love how simple it is to set up and use. The QR login was the one real pain point for me, which is why I care so much about getting it right. I know this PR is still a bit messy and unstructured, and I'm doing my best to bring it into good shape.
> 
> I've been running this branch in my own production setup at home. That's how most of the bugs fixed in the latest months, and it's also why I kept merging upstream into the branch. But I've realised that this is no substitute for proper tests. I've now moved my home deployment to a separate branch, so this PR can stay quiet — I'll stop pushing upstream merges into it. I'll keep collecting and fixing QR-related bugs, but I'll bundle them into a single update here instead of many small pushes.
> 
> I considered splitting this PR into smaller pieces to make the review easier, but for now I've kept it together: most of the change is new, self-contained files rather than edits to existing logic.
> 
> In total the PR touches 46 files (~2,000 lines added, ~200 removed). Excluding the lockfile and translation files, roughly 1,900 added / 140 removed are review-relevant. Please let me know if there's anything I can help with.
> 
> ## Changes overview
> I tried to collect all the changes (with some help from AI) to give a clearer overview, grouped by area.
> 
> ### Backend — self-login client & device flow extensions
> * `service/oidc_service.go`
>   
>   * new `ensureSelfLoginClient` (creates/maintains the internal `pocket-id-self-login` public client)
>   * new `checkDeviceCodePollRate` (RFC 8628 polling interval + `slow_down`)
>   * new `ExchangeDeviceTokenForSession` (turns an authorized device code into a browser session)
>   * small edits to `CreateDeviceAuthorization` / `VerifyDeviceCode`
> * `controller/oidc_controller.go`
>   
>   * new `requireQrLogin` gate
>   * new `handleDeviceFlowError`
>   * new `exchange-session` handler + wiring of the device endpoints
> * `utils/string_util.go`
>   
>   * new `GenerateRandomUnambiguousUppercaseString` for TV-readable user codes (uppercase, no ambiguous characters)
> * `job/db_cleanup_job.go`
>   
>   * cleanup job for expired device codes
> * migrations `*_device_code_rate_limiting.*`
>   
>   * adds `last_polled_at` / `interval_seconds` columns
> * config plumbing for a `qrLoginEnabled` admin toggle
>   
>   * `model`, `dto`, `app_config_service`, `app-config-general-form.svelte`
> 
> ### Frontend — SvelteKit QR-login UI
> grafik
> * `components/qr-login-flow.svelte`
>   
>   * QR code + user code display, countdown, polling and session exchange
> * `components/code-input.svelte`
>   
>   * segmented code input (paste and keyboard navigation, mobile-friendly)
> * `settings/account/device-verification-modal.svelte`
>   
>   * enter/confirm a code coming from another device
> * `utils/device-detect-util.ts`
>   
>   * capability + user-agent detection, routing to the right login page (modern vs. simple)
> * `utils/redirection-util.ts`
>   
>   * shared `isSafeRedirect` helper
> * `services/oidc-service.ts` / `types/oidc.type.ts`
>   
>   * client API methods and types for the device/self-login endpoints
> * route changes
>   
>   * `authorize`, `login`, `login/alternative`: redirect limited devices / no-WebAuthn to the QR or alternative login
>   * `device`, `settings/account`: code-entry entry points
> 
> ### Frontend — standalone "simple" page for legacy browsers
> A separate, non-SvelteKit bundle (no oklch, no Tailwind, ES2015 target) for browsers that can't render the modern UI.
> 
> grafik
> * `simple/src/App.svelte`
>   
>   * the whole standalone app: device-code request, QR rendering, polling, countdown, all states
> * `static/simple/qr/style.css`
>   
>   * hand-written CSS, hardcoded hex colors, webkit-prefixed flexbox
> * `static/simple/qr/index.html`
>   
>   * minimal HTML shell
> * build setup
>   
>   * `vite.config.ts`, `tsconfig.json`, `main.ts`, `qrcode-entry.js`, `build-simple.sh` (IIFE build, bundles the QR library)
> 
> ## Review guide
> A short description of how I'd go through this PR:
> 
> 1. **Backend core** — `oidc_service.go` (`ensureSelfLoginClient`, `checkDeviceCodePollRate`, `ExchangeDeviceTokenForSession`) → `oidc_controller.go` (`exchange-session`, `requireQrLogin`). This is the security-relevant heart.
> 2. **Routing / detection** — `device-detect-util.ts` + `redirection-util.ts`, then the route guards in `authorize/+page.svelte` and `login/+page.svelte`.
> 3. **SvelteKit UI** — `qr-login-flow.svelte`, `code-input.svelte`, `device-verification-modal.svelte`.
> 4. **Standalone page** — `simple/src/App.svelte` + `style.css`. Self-contained, own build, no shared imports.
> 
> ## Bug fixes in the latest push
> * **reauth bypass:** a `reauth=1` URL marker used to persist, so a reload could skip required re-authentication. It's now only honored on limited devices and stripped from the URL after use.
> * **double submit:** the code modal could fire a second request on a repeated Enter; added a guard.
> * **cancel dead-end:** removed a `clearUser()` that left the session unrecoverable when the user cancelled.
> * **odd-length separator:** fixed the code-input separator placement for odd lengths.
> * **state naming:** renamed a `state` variable that collided with the `$state` rune. This also surfaced and fixed a bug where the scope-confirm view went blank during the request.
> * **mobile keyboard:** reworked the code input from one `<input>` per box to a single hidden input behind the boxes, so the on-screen keyboard no longer closes after every character on mobile.
> 
> ## Testing
> Full disclosure: this PR doesn't include automated tests for the new self-login/QR flow yet. I've been running the branch in my own production setup at home, but I know that isn't a substitute for committed tests. I haven't worked much with automated testing in open-source projects so far, but I'd like to learn and add proper tests for this PR. If you have a preference on scope or where they should live, I'll follow it.
> 
> ## AI usage
> In an earlier discussion I mentioned that I'd be using Claude. Just to be transparent: I used an LLM to help with the legacy-page CSS and to review the diff for bugs.

Mate this is bullocks! Well done! <3

**Comment by Norrodar** [comment]
With #1520 the device flow moved to the fosite-based `backend/internal/oidc` package, so
this PR's backend extension now targets code that no longer exists on main. Want me to
port it onto the new structure, or review as-is first? I'd rather ask than dump a big
new diff mid-review.

**Comment by stonith404** [comment]
Hey @Norrodar, sorry for taking soo long to review this PR. It looks not bad but what I don't really like is that this flow is coupled to RFC 8628. While both flows are almost the same they are for two different purposes. RFC 8628 is an OAuth 2.0 authorization flow, which we already support. The suggested feature suggests to add a QR-Login to internal authentication of Pocket ID, this has nothing to do with OAuth 2.0. I completely understand it though why you've implemented this way, because the feature request itself is formulated misleading.

Because of the refactoring we did in #1520 and the resulting merge conflicts and the coupling to  RFC 8628 I thought it's more efficient if I implement the PR from scratch. I'm really curious about your thoughts on the new PR #1594 and would appreciate your review on it. 

What I didn't implement yet is the "simple login page" for older devices, I'm going to add this in a separate PR, we can probably take most of the code from this PR.

**Comment by Norrodar** [comment]
> Hey @Norrodar, sorry for taking soo long to review this PR. [...] I don't really like is that this flow is coupled to RFC 8628. [...] I completely understand it though why you've implemented this way, because the feature request itself is formulated misleading.
> 
> Because of the refactoring [...] I implement the PR from scratch. I'm really curious about your thoughts on the new PR #1594 and would appreciate your review on it.
> 
> What I didn't implement yet is the "simple login page" for older devices, I'm going to add this in a separate PR [...].

No worries about the timing, and thanks for the honest assessment — I fully agree with it. Coupling the login to RFC 8628 forced a few things that never felt quite right (e.g. the internal self-login client). After the refactoring, starting from scratch definitely makes sense! I'm happy to take a close look at [#1594](https://github.com/Norrodar/pocket-id/issues/1594) and leave my feedback there — I already took a sneak peek. :)

Great to hear the simple page for older devices is planned as a follow-up — that was my main use case for the QR feature, so feel free to take whatever is useful from this PR.

Thanks for taking the time to review and rebuild this properly. ♥️
